### DISCUSSION TITLE
What is the best way to test functions that use std functions?

### DISCUSSION THREAD

**Discussion by LucienMorey** [nasa/cFS#609] [Q&A] [answered]
Hi,

I have set up an application that's designed to interface with a shell and will call shell commands with popen, fread and pclose. I am not sure what the best way to unit test this case is because there are no stubs to force returns. Is it possible to manually write stubs for these functions? I see this may have been possible at some point since ```CFS_IO_LIB``` does this for socket-related functions like ```bind```. The other idea is to wrap each of these functions in a utility function inside the application so that the utility function can be stubbed out for testing in the main app. This only kicks the issue further down the line since I would need to test that wrapper function for full coverage.

Is there a recommended way to do this?

**Comment by skliper** [comment, marked as answer]
Writing stubs is the suggested pattern.  OSAL does it quite a bit.  See https://github.com/nasa/osal/tree/main/src/unit-test-coverage/ut-stubs.
